src/System.Net.HttpListener/src/System/Net/Managed/WebSockets/WebSocketValidate.Http.Managed.cs

```diff
@@ -0,0 +1,112 @@
+
+            response.StatusCode = (int)HttpStatusCode.SwitchingProtocols; // HTTP 101
+            response.StatusDescription = HttpStatusDescription.Get(HttpStatusCode.SwitchingProtocols);
+
+            HttpResponseStream responseStream = response.OutputStream as HttpResponseStream;
+            
+            // Send websocket handshake headers
+            await responseStream.WriteWebSocketHandshakeHeadersAsync();
+
+            WebSocket webSocket = ManagedWebSocket.CreateFromConnectedStream(context.Connection.ConnectedStream, true, subProtocol, keepAliveInterval, receiveBufferSize, internalBuffer);
+
+            HttpListenerWebSocketContext webSocketContext = new HttpListenerWebSocketContext(
+                                                                request.Url,
+                                                                request.Headers,
+                                                                request.Cookies,
+                                                                context.User,
+                                                                request.IsAuthenticated,
+                                                                request.IsLocal,
+                                                                request.IsSecureConnection,
+                                                                origin,
+                                                                secWebSocketProtocols.AsReadOnly(),
+                                                                secWebSocketVersion,
+                                                                secWebSocketKey,
+                                                                webSocket);
+
+            return webSocketContext;
+        }
+
+        private static void ValidateWebSocketHeaders(HttpListenerContext context)
+        {
+            if (!context.Request.IsWebSocketRequest)
+            {
+                throw new WebSocketException(WebSocketError.NotAWebSocket,
+                    SR.Format(SR.net_WebSockets_AcceptNotAWebSocket,
+                    nameof(ValidateWebSocketHeaders),
+                    HttpKnownHeaderNames.Connection,
+                    HttpKnownHeaderNames.Upgrade,
+                    WebSocketValidate.WebSocketUpgradeToken,
+                    context.Request.Headers[HttpKnownHeaderNames.Upgrade]));
+            }
+
+            string secWebSocketVersion = context.Request.Headers[HttpKnownHeaderNames.SecWebSocketVersion];
+            if (string.IsNullOrEmpty(secWebSocketVersion))
+            {
+                throw new WebSocketException(WebSocketError.HeaderError,
+                    SR.Format(SR.net_WebSockets_AcceptHeaderNotFound,
+                    nameof(ValidateWebSocketHeaders),
+                    HttpKnownHeaderNames.SecWebSocketVersion));
+            }
+
+            // TODO: Figure out how to validate version.
```

Comment:
What is the concern here?  If this remains a TODO, an issue should be opened for it.